File: src/main/java/com/tkr/desk/OrderBlotter.java
package com.tkr.desk;

import com.tkr.types.WireTypes.BatchWireFrame;
import com.tkr.types.WireTypes.Status;
import com.tkr.types.WireTypes.WireBatchRecord;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Maintains an in-memory order blotter for allocation batches awaiting desk routing.
 */
public final class OrderBlotter {

    public static final class BlotterEntry {
        public int entryId;
        public int batchId;
        public int recordId;
        public int accountId;
        public int symbolId;
        public int qtyMilli;
        public int priceTick;
        public long notionalCents;
        public long createdAtMillis;
        public String routeState;
    }

    public static final class BlotterSnapshot {
        public Status status = Status.OK;
        public List<BlotterEntry> entries = new ArrayList<>();
        public long totalNotionalCents;
        public int openOrders;
    }

    private final Map<Integer, BlotterEntry> openByRecordId = new HashMap<>();
    private int nextEntryId = 1;
    private final long createdBaselineMillis;

    public OrderBlotter(long createdBaselineMillis) {
        this.createdBaselineMillis = createdBaselineMillis;
    }

    public Status ingestBatch(BatchWireFrame frame) {
        if (frame == null || frame.records == null) {
            return Status.BOUNDS_ERROR;
        }
        long ts = createdBaselineMillis + frame.header.tradeDateYyyymmdd;
        for (WireBatchRecord rec : frame.records) {
            BlotterEntry entry = new BlotterEntry();
            entry.entryId = nextEntryId++;
            entry.batchId = frame.header.deskId;
            entry.recordId = rec.recordId;
            entry.accountId = rec.accountId;
            entry.symbolId = rec.symbolId;
            entry.qtyMilli = rec.qtyMilli;
            entry.priceTick = rec.priceTick;
            entry.notionalCents = (long) rec.qtyMilli * rec.priceTick / 1000L;
            entry.createdAtMillis = ts + rec.recordId;
            entry.routeState = "OPEN";
            openByRecordId.put(rec.recordId, entry);
        }
        return Status.OK;
    }

    public BlotterSnapshot snapshot() {
        BlotterSnapshot snap = new BlotterSnapshot();
        snap.entries.addAll(openByRecordId.values());
        snap.entries.sort(Comparator.comparingInt(e -> e.entryId));
        for (BlotterEntry entry : snap.entries) {
            snap.totalNotionalCents += entry.notionalCents;
            if ("OPEN".equals(entry.routeState)) {
                snap.openOrders++;
            }
        }
        return snap;
    }

    public Status markRouted(int recordId) {
        BlotterEntry entry = openByRecordId.get(recordId);
        if (entry == null) {
            return Status.BOUNDS_ERROR;
        }
        entry.routeState = "ROUTED";
        return Status.OK;
    }

    public Status markFilled(int recordId, int filledQtyMilli) {
        BlotterEntry entry = openByRecordId.get(recordId);
        if (entry == null) {
            return Status.BOUNDS_ERROR;
        }
        if (filledQtyMilli <= 0 || filledQtyMilli > entry.qtyMilli) {
            return Status.BOUNDS_ERROR;
        }
        entry.qtyMilli -= filledQtyMilli;
        entry.notionalCents = (long) entry.qtyMilli * entry.priceTick / 1000L;
        entry.routeState = entry.qtyMilli == 0 ? "FILLED" : "PARTIAL";
        if (entry.qtyMilli == 0) {
            openByRecordId.remove(recordId);
        }
        return Status.OK;
    }

    public List<BlotterEntry> entriesForAccount(int accountId) {
        List<BlotterEntry> out = new ArrayList<>();
        for (BlotterEntry entry : openByRecordId.values()) {
            if (entry.accountId == accountId) {
                out.add(entry);
            }
        }
        out.sort(Comparator.comparingInt(e -> e.symbolId));
        return out;
    }

    public long openNotionalForSymbol(int symbolId) {
        long total = 0;
        for (BlotterEntry entry : openByRecordId.values()) {
            if (entry.symbolId == symbolId && "OPEN".equals(entry.routeState)) {
                total += entry.notionalCents;
            }
        }
        return total;
    }

    public void clearFilled() {
        openByRecordId.entrySet().removeIf(e -> "FILLED".equals(e.getValue().routeState));
    }

    public int openCount() {
        return openByRecordId.size();
    }
}

File: legacy-cpp/src/desk/order_blotter.cc
#include "tkr/desk/order_blotter.h"

#include <algorithm>

namespace tkr {
namespace desk {

OrderBlotter::OrderBlotter(std::int64_t created_baseline_millis)
    : next_entry_id_(1), created_baseline_millis_(created_baseline_millis) {}

Status OrderBlotter::IngestBatch(const BatchWireFrame& frame) {
  const std::int64_t ts =
      created_baseline_millis_ + static_cast<std::int64_t>(frame.header.trade_date_yyyymmdd);
  for (const WireBatchRecord& rec : frame.records) {
    BlotterEntry entry{};
    entry.entry_id = next_entry_id_++;
    entry.batch_id = frame.header.desk_id;
    entry.record_id = rec.record_id;
    entry.account_id = rec.account_id;
    entry.symbol_id = rec.symbol_id;
    entry.qty_milli = rec.qty_milli;
    entry.price_tick = rec.price_tick;
    entry.notional_cents =
        (static_cast<std::int64_t>(rec.qty_milli) *
         static_cast<std::int64_t>(rec.price_tick)) /
        1000;
    entry.created_at_millis = ts + rec.record_id;
    entry.route_state = "OPEN";
    open_by_record_id_[rec.record_id] = entry;
  }
  return Status::kOk;
}

BlotterSnapshot OrderBlotter::Snapshot() const {
  BlotterSnapshot snap{};
  snap.status = Status::kOk;
  snap.entries.reserve(open_by_record_id_.size());
  for (const auto& entry : open_by_record_id_) {
    snap.entries.push_back(entry.second);
  }
  std::sort(snap.entries.begin(), snap.entries.end(),
            [](const BlotterEntry& a, const BlotterEntry& b) {
              return a.entry_id < b.entry_id;
            });
  for (const BlotterEntry& entry : snap.entries) {
    snap.total_notional_cents += entry.notional_cents;
    if (entry.route_state == "OPEN") {
      ++snap.open_orders;
    }
  }
  return snap;
}

Status OrderBlotter::MarkRouted(std::uint32_t record_id) {
  auto it = open_by_record_id_.find(record_id);
  if (it == open_by_record_id_.end()) {
    return Status::kBoundsError;
  }
  it->second.route_state = "ROUTED";
  return Status::kOk;
}

Status OrderBlotter::MarkFilled(std::uint32_t record_id,
                                std::uint32_t filled_qty_milli) {
  auto it = open_by_record_id_.find(record_id);
  if (it == open_by_record_id_.end()) {
    return Status::kBoundsError;
  }
  BlotterEntry& entry = it->second;
  if (filled_qty_milli <= 0 || filled_qty_milli > entry.qty_milli) {
    return Status::kBoundsError;
  }
  entry.qty_milli -= filled_qty_milli;
  entry.notional_cents =
      (static_cast<std::int64_t>(entry.qty_milli) *
       static_cast<std::int64_t>(entry.price_tick)) /
      1000;
  entry.route_state = entry.qty_milli == 0 ? "FILLED" : "PARTIAL";
  if (entry.qty_milli == 0) {
    open_by_record_id_.erase(it);
  }
  return Status::kOk;
}

std::vector<BlotterEntry> OrderBlotter::EntriesForAccount(
    std::uint32_t account_id) const {
  std::vector<BlotterEntry> out;
  for (const auto& entry : open_by_record_id_) {
    if (entry.second.account_id == account_id) {
      out.push_back(entry.second);
    }
  }
  std::sort(out.begin(), out.end(),
            [](const BlotterEntry& a, const BlotterEntry& b) {
              return a.symbol_id < b.symbol_id;
            });
  return out;
}

std::int64_t OrderBlotter::OpenNotionalForSymbol(std::uint32_t symbol_id) const {
  std::int64_t total = 0;
  for (const auto& entry : open_by_record_id_) {
    if (entry.second.symbol_id == symbol_id && entry.second.route_state == "OPEN") {
      total += entry.second.notional_cents;
    }
  }
  return total;
}

void OrderBlotter::ClearFilled() {
  for (auto it = open_by_record_id_.begin(); it != open_by_record_id_.end();) {
    if (it->second.route_state == "FILLED") {
      it = open_by_record_id_.erase(it);
    } else {
      ++it;
    }
  }
}

std::int32_t OrderBlotter::OpenCount() const {
  return static_cast<std::int32_t>(open_by_record_id_.size());
}

}  // namespace desk
}  // namespace tkr
